db/dbtypes/types.go

```diff
@@ -939,27 +939,25 @@ func CountMergedRowsCompact(rows []AddressRowCompact, txnView AddrTxnViewType) (
 // positive or not, although the Value field is an absolute value (always
 // positive). MergedRows will return a non-nil error of a merged row is detected
 // in the input since only non-merged rows are expected.
-func MergeRows(rows []*AddressRow) ([]AddressRowMerged, error) {
-	hashes := make([]chainhash.Hash, 0, len(rows))
-	merged := make(map[chainhash.Hash]*AddressRowMerged)
+func MergeRows(rows []*AddressRow) ([]*AddressRowMerged, error) {
+	numUniqueHashesGuess := len(rows)*2/3 + 1
```

Comment:
The number of unique transaction hashes is not known from the input address rows since an address may have multiple appearances in a given transaction.  Still pre-allocate, since we have an idea of the ballpark size of the result, but try not to overshoot.

Reply:
Any particular reasoning behind this guess?

Reply:
The actual number could be smaller or even larger than in input number of address rows.  Since we don't know what it is, I decided not to over allocate space, which is wasted.  On the other hand, when go needs to append to a slice with insufficient capacity, it automatically expands the capacity by re-allocating (and copying the elements to the new larger buffer), which not free.  As for the exact fraction, it's more or less pulled out of thin air.  But when go expands capacity to append, it doubles capacity until it hits 1024, after which further appends will cause a 25% capacity increase.  Considering a large slice, of say 409,600 pointers, where it's more costly to reallocate, it is best not to start with a zero-capacity slice as it would involve 36 reallocations and 14,632,472 bytes copied in the process.  https://play.golang.org/p/wZTdS7TqjX3
So `2/3` is just balancing the desire to avoid reallocation and copying, with the desire to not waste space when the actual output length is much less than the input length (which does happen, even in our `insight-addr` attack profile.
I definitely do not think I picked the optimal number here, so I'm open to other ideas.  One thought I had was to overallocate and when the loop finishes, make a new slice with the correct capacity.